File: app/services/rest_sender.py
"""REST sending seam — the ONE place that turns a prepared request into a result.

Two strategies sit behind a single :func:`send_request` contract:

* **mock** (``tls is None``) — no network call: echoes the caller's editable
  ``mock_response`` after a small simulated latency, exactly as the old stub did.
  Used when a template has no configured preset connection.
* **real** (``tls`` given) — an actual HTTPS request via ``httpx`` using the
  browser-provided client certificate (PEM cert+key, or a JKS keystore). Used
  when the template's preset has a connection configured. The certificate lives
  only in the browser (sessionStorage) and arrives on this one call.

The client-cert material is held only in memory + short-lived ``0600`` temp files
that are deleted before returning; it is never persisted or logged. Keep this
module the single seam: swapping the real branch for an external "send for us"
API later means editing only :func:`_send_real`.
"""

from __future__ import annotations

import asyncio
import base64
import contextlib
import os
import random
import ssl
import tempfile
import textwrap
import time
from collections.abc import Iterator
from dataclasses import dataclass
from typing import Any

import httpx

from app.utils.status_code import extract_status_code

# The mock's response headers/status — kept identical to the old stub so the
# history rows and UI badges are unchanged when no connection is configured.
_MOCK_RESPONSE_HEADERS = {"Content-Type": "application/json; charset=utf-8", "X-Mock-Send": "true"}
_REQUEST_TIMEOUT = 30.0


@dataclass
class TlsMaterial:
    """Client-certificate material for one real send, provided by the browser.

    Either PEM (``cert_pem`` + ``key_pem``, key optionally encrypted with
    ``password``) or a JKS keystore (``jks_b64`` + ``password``). ``verify``
    mirrors the per-preset "проверять сертификат сервера" toggle (default off so
    self-signed test endpoints work).
    """

    kind: str  # "pem" | "jks"
    cert_pem: str = ""
    key_pem: str = ""
    jks_b64: str = ""
    password: str = ""
    verify: bool = False


@dataclass
class SendResult:
    """Outcome of a send, shaped for both the JSON response and history record."""

    ok: bool
    http_status: int | None
    status_code: int | None
    response_headers: dict[str, str]
    response_body: str
    latency_ms: int | None
    status_text: str = "OK"
    error_message: str = ""


def tls_from_payload(payload: dict[str, Any]) -> TlsMaterial | None:
    """Build :class:`TlsMaterial` from a send payload's optional ``tls`` block.

    Returns ``None`` (→ mock send) when no usable connection is present: no
    ``tls`` object, an unknown ``kind``, or missing required material. The
    browser only sends a complete block, so ``None`` reliably means "no
    configured preset connection".
    """

    raw = payload.get("tls")
    if not isinstance(raw, dict):
        return None
    kind = str(raw.get("kind") or "").strip().lower()
    verify = bool(raw.get("verify"))
    password = str(raw.get("password") or "")
    if kind == "pem":
        cert_pem = str(raw.get("cert") or "")
        key_pem = str(raw.get("key") or "")
        if not cert_pem or not key_pem:
            return None
        return TlsMaterial(
            kind="pem", cert_pem=cert_pem, key_pem=key_pem, password=password, verify=verify
        )
    if kind == "jks":
        jks_b64 = str(raw.get("jks") or "")
        if not jks_b64:
            return None
        return TlsMaterial(kind="jks", jks_b64=jks_b64, password=password, verify=verify)
    return None


async def send_request(
    *,
    method: str,
    url: str,
    headers: Any,
    body: str,
    mock_response: str,
    tls: TlsMaterial | None,
) -> SendResult:
    """Send one request. ``tls is None`` → mock; otherwise a real mTLS send."""

    if tls is None:
        return await _send_mock(mock_response)
    return await _send_real(method=method, url=url, headers=headers, body=body, tls=tls)


# ---------- mock strategy ----------


async def _simulate_latency(latency_ms: int) -> None:
    """Mock network latency — module-local so tests can patch it."""

    await asyncio.sleep(latency_ms / 1000)


async def _send_mock(mock_response: str) -> SendResult:
    latency_ms = random.randint(35, 220)
    await _simulate_latency(latency_ms)
    status_code = extract_status_code(mock_response)
    # Transport never fails for the mock; a non-zero business statusCode in the
    # body is the only failure signal (shown red in the UI), absent/zero = ok.
    ok = status_code is None or status_code == 0
    return SendResult(
        ok=ok,
        http_status=200,
        status_code=status_code,
        response_headers=dict(_MOCK_RESPONSE_HEADERS),
        response_body=mock_response,
        latency_ms=latency_ms,
    )


# ---------- real strategy (httpx + client cert) ----------


async def _send_real(*, method: str, url: str, headers: Any, body: str, tls: TlsMaterial) -> SendResult:
    started = time.perf_counter()
    try:
        with _client_ssl_context(tls) as ctx:
            async with httpx.AsyncClient(verify=ctx, timeout=_REQUEST_TIMEOUT) as client:
                resp = await client.request(
                    (method or "GET").upper(),
                    url,
                    headers=_header_pairs(headers),
                    content=body.encode("utf-8") if body else None,
                )
        latency_ms = int((time.perf_counter() - started) * 1000)
        response_body = resp.text
        status_code = extract_status_code(response_body)
        # Overall ok = transport succeeded (2xx/3xx) AND no non-zero business
        # statusCode in the body — matching the mock's business-code semantics.
        ok = 200 <= resp.status_code < 400 and (status_code is None or status_code == 0)
        return SendResult(
            ok=ok,
            http_status=resp.status_code,
            status_code=status_code,
            response_headers=dict(resp.headers),
            response_body=response_body,
            latency_ms=latency_ms,
            status_text=resp.reason_phrase or "",
        )
    except Exception as exc:  # noqa: BLE001 — network / TLS / cert-parse all map to a failed send
        latency_ms = int((time.perf_counter() - started) * 1000)
        return SendResult(
            ok=False,
            http_status=None,
            status_code=None,
            response_headers={},
            response_body="",
            latency_ms=latency_ms,
            status_text="",
            error_message=_error_text(exc),
        )


def _header_pairs(headers: Any) -> list[tuple[str, str]]:
    """Flatten the resolved header list ({key,value,disabled,…}) to (k, v) pairs,
    skipping disabled/blank rows. Never includes any TLS material."""

    out: list[tuple[str, str]] = []
    if isinstance(headers, list):
        for h in headers:
            if not isinstance(h, dict) or h.get("disabled"):
                continue
            key = str(h.get("key") or "").strip()
            if key:
                out.append((key, str(h.get("value") or "")))
    return out


@contextlib.contextmanager
def _client_ssl_context(tls: TlsMaterial) -> Iterator[ssl.SSLContext]:
    """Yield an SSLContext carrying the client cert; delete temp files on exit."""

    cert_pem, key_pem, key_password = _materialize_pem(tls)
    tmp_paths: list[str] = []
    try:
        cert_file = _write_secret(cert_pem, tmp_paths)
        key_file = _write_secret(key_pem, tmp_paths)
        ctx = ssl.create_default_context()
        if not tls.verify:
            # Order matters: disable hostname check before dropping verify_mode,
            # else create_default_context() raises.
            ctx.check_hostname = False
            ctx.verify_mode = ssl.CERT_NONE
        ctx.load_cert_chain(certfile=cert_file, keyfile=key_file, password=key_password)
        yield ctx
    finally:
        for path in tmp_paths:
            with contextlib.suppress(OSError):
                os.remove(path)


def _materialize_pem(tls: TlsMaterial) -> tuple[str, str, str | None]:
    """Return (cert_pem, key_pem, key_password). JKS is decoded to PEM here; its
    extracted key is already decrypted, so no password is passed to OpenSSL."""

    if tls.kind == "jks":
        cert_pem, key_pem = _jks_to_pem(tls.jks_b64, tls.password)
        return cert_pem, key_pem, None
    return tls.cert_pem, tls.key_pem, (tls.password or None)


def _jks_to_pem(jks_b64: str, password: str) -> tuple[str, str]:
    """Extract the first private key + its cert chain from a JKS as PEM text."""

    import jks  # pyjks — imported lazily so the dep is only needed for JKS

    store = jks.KeyStore.loads(base64.b64decode(jks_b64), password)
    for entry in store.private_keys.values():
        if not entry.is_decrypted():
            entry.decrypt(password)
        key_pem = _der_to_pem(entry.pkey_pkcs8, "PRIVATE KEY")
        cert_pem = "".join(_der_to_pem(cert, "CERTIFICATE") for _, cert in entry.cert_chain)
        return cert_pem, key_pem
    raise ValueError("В JKS нет приватного ключа")


def _der_to_pem(der: bytes, label: str) -> str:
    body = "\n".join(textwrap.wrap(base64.b64encode(der).decode("ascii"), 64))
    return f"-----BEGIN {label}-----\n{body}\n-----END {label}-----\n"


def _write_secret(pem: str, tmp_paths: list[str]) -> str:
    """Write PEM text to a fresh 0600 temp file, tracked for later deletion."""

    fd, path = tempfile.mkstemp(suffix=".pem")
    tmp_paths.append(path)
    os.chmod(path, 0o600)
    with os.fdopen(fd, "w") as fh:
        fh.write(pem)
    return path


def _error_text(exc: Exception) -> str:
    text = str(exc).strip()
    return text or exc.__class__.__name__

File: app/static/js/rest_sender.js
/*
 * Browser-side REST sending seam — the ONE place that turns a prepared request
 * into a result. Mirrors the retired server seam (app/services/rest_sender.py):
 *
 *  - mock (real=false): no network call — echoes the caller's editable
 *    mock_response after a small simulated latency, exactly as the old stub.
 *  - real (real=true): a direct fetch() from THIS browser to the target URL.
 *    Сообщение не проходит через наш бэкенд; клиентский сертификат (mTLS)
 *    браузер берёт из системного хранилища ОС, серверный сертификат всегда
 *    проверяется браузером (отключить это из JS невозможно).
 *
 * Result shape is the same JSON the old /send-htmx/execute endpoint returned,
 * so panel code keeps reading d.ok / d.status / d.body / d.error unchanged:
 *   { ok, status, status_text, latency_ms, headers, body, error }
 *
 * Loaded (defer) after status_code.js (needs window.extractStatusCode).
 */
(function () {
    var REQUEST_TIMEOUT_MS = 30000;

    // The mock's response headers/status — kept identical to the old server stub
    // so history rows and UI badges are unchanged when real sending is off.
    var MOCK_RESPONSE_HEADERS = {
        'Content-Type': 'application/json; charset=utf-8',
        'X-Mock-Send': 'true',
    };

    // Flatten resolved header rows ({key,value,disabled?}) to [key, value]
    // pairs, skipping disabled/blank-key rows. A legitimate falsy value (e.g.
    // "0") must survive verbatim — mirror of the server's _header_pairs.
    function headerPairs(rows) {
        var out = [];
        (Array.isArray(rows) ? rows : []).forEach(function (h) {
            if (!h || typeof h !== 'object' || h.disabled) return;
            var key = String(h.key == null ? '' : h.key).trim();
            if (!key) return;
            out.push([key, h.value == null ? '' : String(h.value)]);
        });
        return out;
    }

    // Overall ok = transport succeeded (2xx) AND no non-zero business statusCode
    // in the body — the same semantics for mock and real.
    function computeOk(httpStatus, statusCode) {
        var transport = typeof httpStatus === 'number' && httpStatus >= 200 && httpStatus < 300;
        return transport && (statusCode === null || statusCode === 0);
    }

    // Human-readable (Russian) text for a failed fetch. The browser hides the
    // real cause of a network-layer failure (opaque TypeError), so the CORS/
    // network/cert possibilities are listed instead of guessed.
    function errorText(err, url, pageProtocol) {
        if (err && err.name === 'AbortError') {
            return 'Сервис не ответил за ' + (REQUEST_TIMEOUT_MS / 1000) + ' с — отправка прервана по таймауту.';
        }
        if (pageProtocol === 'https:' && /^http:\/\//i.test(String(url))) {
            return 'Браузер блокирует запрос на http:// со страницы https:// (mixed content). Используйте https-адрес сервиса.';
        }
        return 'Запрос не ушёл из браузера. Возможные причины: '
            + 'сервис не разрешает CORS-запросы с этого адреса; '
            + 'нет сетевого доступа до хоста; '
            + 'сервис требует клиентский сертификат, который не установлен в системе; '
            + 'самоподписанный сертификат сервера не является доверенным.';
    }

    function extractStatusCode(body) {
        return typeof window.extractStatusCode === 'function' ? window.extractStatusCode(body) : null;
    }

    function result(fields) {
        return {
            ok: !!fields.ok,
            status: fields.status != null ? fields.status : null,
            status_text: fields.status_text || '',
            latency_ms: fields.latency_ms != null ? fields.latency_ms : null,
            headers: fields.headers || {},
            body: fields.body || '',
            error: fields.error || '',
        };
    }

    // ---------- mock strategy ----------

    function sendMock(req, deps) {
        var latency = 35 + Math.floor(deps.random() * 186); // 35..220 ms, as the old stub
        return deps.sleep(latency).then(function () {
            var body = String(req.mockResponse == null ? '' : req.mockResponse);
            var statusCode = extractStatusCode(body);
            return result({
                ok: statusCode === null || statusCode === 0,
                status: 200,
                status_text: 'OK',
                latency_ms: latency,
                headers: Object.assign({}, MOCK_RESPONSE_HEADERS),
                body: body,
            });
        });
    }

    // ---------- real strategy (direct fetch from the browser) ----------

    function sendReal(req, deps) {
        var method = String(req.method || 'GET').toUpperCase();
        var pairs = headerPairs(req.headers);
        var headers = {};
        var hasContentType = false;
        pairs.forEach(function (kv) {
            if (kv[0].toLowerCase() === 'content-type') hasContentType = true;
            headers[kv[0]] = kv[1];
        });
        var body = String(req.body == null ? '' : req.body);
        // fetch() rejects a GET/HEAD with a body — drop it (httpx allowed it, but
        // the app is POST-centric; the envelope's method is what matters).
        var withBody = body !== '' && method !== 'GET' && method !== 'HEAD';
        // httpx sent no Content-Type at all; fetch would force text/plain, which
        // CORS-preflights differently and confuses JSON services — default it
        // from the template's format instead.
        if (withBody && !hasContentType) {
            headers['Content-Type'] = req.format === 'json' ? 'application/json' : 'application/xml';
        }
        var controller = typeof AbortController !== 'undefined' ? new AbortController() : null;
        var timer = controller ? setTimeout(function () { controller.abort(); }, REQUEST_TIMEOUT_MS) : null;
        var started = deps.now();

        function elapsed() { return Math.round(deps.now() - started); }

        return deps.fetchImpl(String(req.url || ''), {
            method: method,
            headers: headers,
            body: withBody ? body : undefined,
            mode: 'cors',
            // Don't follow redirects — the old server seam treated a 3xx as an
            // unhandled (not-ok) response, not a success. The browser hides the
            // numeric 3xx code behind an opaqueredirect.
            redirect: 'manual',
            signal: controller ? controller.signal : undefined,
        }).then(function (r) {
            if (r.type === 'opaqueredirect') {
                return result({
                    ok: false,
                    latency_ms: elapsed(),
                    error: 'Сервис ответил перенаправлением (3xx); оно не выполняется, точный код скрыт браузером.',
                });
            }
            return r.text().then(function (text) {
                var respHeaders = {};
                r.headers.forEach(function (value, key) { respHeaders[key] = value; });
                var statusCode = extractStatusCode(text);
                return result({
                    ok: computeOk(r.status, statusCode),
                    status: r.status,
                    status_text: r.statusText || '',
                    latency_ms: elapsed(),
                    headers: respHeaders,
                    body: text,
                });
            });
        }).catch(function (err) {
            var pageProtocol = (typeof location !== 'undefined' && location.protocol) || '';
            return result({
                ok: false,
                latency_ms: elapsed(),
                error: errorText(err, req.url, pageProtocol),
            });
        }).finally(function () {
            if (timer) clearTimeout(timer);
        });
    }

    // ---------- public seam ----------

    window.restSender = {
        // send({method,url,headers,body,format,mockResponse,real}) -> Promise<result>
        // `deps` is test-only injection: {fetchImpl, sleep, random, now}.
        send: function (req, deps) {
            var d = deps || {};
            var full = {
                fetchImpl: d.fetchImpl || function (u, opts) { return fetch(u, opts); },
                sleep: d.sleep || function (ms) { return new Promise(function (res) { setTimeout(res, ms); }); },
                random: d.random || Math.random,
                now: d.now || function () {
                    return (typeof performance !== 'undefined' && performance.now) ? performance.now() : Date.now();
                },
            };
            return req && req.real ? sendReal(req, full) : sendMock(req, full);
        },
    };

    // Node-only export so the pure logic can be unit-tested without a browser
    // (see tests/js/rest_sender.test.js). No-op in the browser.
    if (typeof module !== 'undefined' && module.exports) {
        module.exports = {
            restSender: window.restSender,
            headerPairs: headerPairs,
            computeOk: computeOk,
            errorText: errorText,
        };
    }
})();
